components/PlaybackTask.brs

```diff
@@ -0,0 +1,253 @@
+  videoPlayer.EnableCookies()
+
+  m.videoPlayer = videoPlayer
+end sub
+
+sub setupRaf()
+  ? "TRUE[X] >>> setupRaf()"
+  raf = Roku_Ads()
+  raf.enableAdMeasurements(true)
+  raf.setContentGenre("Entertainment")
+  raf.setContentId("TrueXSample")
+
+  raf.SetDebugOutput(false) 'debugging
+  raf.SetAdPrefs(false)
+
+  adUrl = m.top.adUrl
+  if adUrl = invalid OR adUrl = ""
+    ' adUrl = "pkg:/res/adpods/vast-regular.xml"  ' Regular Ads
+    adUrl = "pkg:/res/adpods/vmap-truex.xml" ' Preroll + Midroll Truex experience
+    ' adUrl = "pkg:/res/adpod/truex-pod-preroll.xml"
+    ' adUrl = "pkg:/res/adpod/truex-pod-skip.xml" ' Forced standalone skip pod.  Can be inserted in a full ad payload
+  end if 
+  raf.setAdUrl(adUrl)
+
+  m.raf = raf
+end sub
+
+sub setupEvents()
+  m.top.observeField("exitPlayback", m.port)
+end sub
+
+sub initPlayback()
+  ? "TRUE[X] >>> initPlayback()"
+
+  m.currentAdPod = getPreroll()
+  startPlayback() ' Will handle playing a preroll if it exists per above
+
+  while(true)
+    msg = Wait(0, m.port)
+    msgType = type(msg)
+
+    ' ? "event type:" msgType
+    if msgType = "roSGNodeEvent"
+      field = msg.getField()
+      ' ? "roSGNodeEvent msg.getField()" field
+      if field = "position" then 
+        onPositionChanged(msg)
+      else if field = "event" then
+        onTrueXEvent(msg)
+      else if field = "exitPlayback"
+        exitPlayback()
+      end if
+      ' TODO: Video complete event
+    end if
+  end while
+end sub
+
+sub onPositionChanged(event)
+  m.lastPosition = event.getData()
+
+  ads = m.raf.getAds(event)
+  handleAds(ads)
+end sub
+
+sub onTruexEvent(event)
+  data = event.getData()
+  eventType = data.type
+
+  ? "TRUE[X] >>> PlaybackTask::onTrueXEvent(): " eventType
+
+  types = {
+    "ADSTARTED": "adStarted",
+    "ADCOMPLETED": "adCompleted"
+    "ADPOSITION": "adPosition",
+    "ADERROR": "adError",
+    "ADFREEPOD": "adFreePod"
+    "NOADSAVAILABLE": "noAdsAvailable",
+    "OPTIN": "optIn",
+    "OPTOUT": "optOut",
+    "USERCANCEL": "userCancel",
+    "USERCANCELSTREAM": "userCancelStream",
+    "SKIPCARDSHOWN": "skipCardShown"
+  }
+
+  startPlaybackEvents = [types.ADCOMPLETED, types.NOADSAVAILABLE, types.ADERROR]
+
+  if eventType = types.ADFREEPOD
+    m.skipAds = true
+  else if arrayUtils_includes(startPlaybackEvents, eventType)
+    startPlayback()
+  else if eventType = types.USERCANCELSTREAM
+    exitPlayback()
+  end if
+end sub
+
+function getPreroll() as Object
+  ads = m.raf.getAds()
+  if ads = invalid then return false
+  
+  result = invalid
+  for each adPod in ads
+    if adPod.rendersequence = "preroll"
+      result = adPod
+      exit for
+    end if
+  end for
+
+  return result
+end function
+
+function handleAds(ads) as Boolean
+  resumePlayback = true
+
+  if ads <> invalid AND ads.ads.count() > 0
+    m.currentAdPod = ads
+    firstAd = ads.ads[0] 'Assume truex can only be first ad
+
+    if firstAd.adParameters <> invalid
```

Comment:
True, the ads might not be from TrueX but use the same tags.  Let me look at whats returned by RAF here more closely.

Reply:
`AdSystem` is not parsed by RAF unfortunately.

I opted for `adServer` property, to avoid having to dive into parsing the adparameters and figure that TAR will reject bad urls.  Assuming that all TrueX ad's will be sourced by `get.truex.com`.  For simplicity I opted not to use a regex.

Reply:
you should support qa-get.truex.com as well

Reply:
Good point.  The code happens to check it right now, but I'll make it more explicit.